### PR TITLE
[JENKINS-26107] Using blocky stages

### PR DESCRIPTION
**Pull request by jglick**
Subsumes #3 by just skipping `milestone` when unavailable, so normally just demonstrates blocky `stage` and `lock`. In practice this is good enough for a demo.

By the way @svanoort I played with running nested stages:

```diff
@@ -53,7 +53,9 @@
     node {
         checkout scm
         servers.runWithServer {id ->
-            mvn "-o -f sometests test -Durl=${jettyUrl}${id}/ -Dduration=${duration}"
+            stage("${duration}s tests") {
+                mvn "-o -f sometests test -Durl=${jettyUrl}${id}/ -Dduration=${duration}"
+            }
         }
         junit '**/target/surefire-reports/TEST-*.xml'
     }
```

It is not too bad; you see time taken in QA setup vs. time taken actually running `mvn test`:

stages

The main quirk is that when a new build starts running, say, the 30s tests, but has not yet even gotten a node to run the 20s tests on, the stage view will *hide* earlier builds, because it refuses to show builds which do not exactly align with the stages shown in the most recent. I believe I filed this in the past. Also the timing information does not take parallelism into account in a sensible fashion.

@reviewbybees

### CODE REVIEW

File: demo/Dockerfile [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
+FROM jenkins:2.7.2
 
 USER root
 
 # installed maven package is still 3.0.5, pretty old
 ENV MAVEN_VERSION 3.3.9
-RUN cd /usr/local; wget -O - http://mirrors.ibiblio.org/apache/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz | tar xvzf -
-RUN ln -sv /usr/local/apache-maven-$MAVEN_VERSION /usr/local/maven
+RUN cd /usr/local; wget -q -O - http://mirrors.ibiblio.org/apache/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz | tar xvfz - && \
+    ln -sv /usr/local/apache-maven-$MAVEN_VERSION /usr/local/maven
```
Comment (jglick) [inline demo/Dockerfile]: Unrelated, just routine `Dockerfile` maintenance.

File: demo/plugins.txt [hunk 2]
```diff
@@ -3,33 +3,35 @@ branch-api:1.10
...
+pipeline-input-step:2.1
+pipeline-rest-api:2.0
+pipeline-stage-step:2.2
+pipeline-stage-view:2.0
 scm-api:1.2
-script-security:1.21
+script-security:1.22
 ssh-credentials:1.12
-structs:1.3
+structs:1.4
```
Comment (jglick) [inline demo/plugins.txt]: Note that I just released 1.5 but it is not yet on the UC. The bug in 1.4 does not seem to affect this demo anyway.

File: demo/plugins.txt [hunk 3]
```diff
@@ -3,33 +3,35 @@ branch-api:1.10
 cloudbees-folder:5.12
 credentials:2.1.4
 durable-task:1.12
-git:2.5.2
-git-client:1.19.7
+git:2.5.3
+git-client:1.21.0
```
Comment (jglick) [inline demo/plugins.txt]: Just updating everything while I am here.

File: demo/repo/Jenkinsfile [hunk 4]
```diff
@@ -50,3 +58,11 @@ def runTests(servers, duration) {
         junit '**/target/surefire-reports/TEST-*.xml'
     }
 }
+
+def milestone(n) {
+    try {
+        steps.milestone(n)
+    } catch (NoSuchMethodError _) {
```
Comment (jglick) [inline demo/repo/Jenkinsfile]: Same trick as for `checkpoint`.

File: .gitignore [hunk 5]
```diff
@@ -1,2 +1,5 @@
 target
 work
+.project
```
Comment (svanoort) [inline .gitignore]: Comment: we should consider applying a standard .gitignore to all our projects (exclude IDE files, build files, work, and JS node stuff).

File: demo/Dockerfile [hunk 6]
```diff
@@ -1,21 +1,21 @@
...
 # install Jetty
 WORKDIR /opt
 # jetty package is still 8
 ENV JETTY_VERSION 9.2.12.v20150709
-RUN wget -O - "http://archive.eclipse.org/jetty/$JETTY_VERSION/dist/jetty-distribution-$JETTY_VERSION.tar.gz" | tar xvfz -
-RUN ln -sv jetty-distribution-$JETTY_VERSION jetty
-RUN cd /tmp; ln -s /opt/jetty/webapps
-RUN chown -R jenkins /opt/jetty/logs
-RUN chown -R jenkins /opt/jetty/webapps
+RUN wget -q -O - "http://archive.eclipse.org/jetty/$JETTY_VERSION/dist/jetty-distribution-$JETTY_VERSION.tar.gz" | tar xvfz - && \
```
Comment (svanoort) [inline demo/Dockerfile]: Huzzah!  Efficient layer use in docker.

File: demo/repo/Jenkinsfile (hunk 4 above)
Comment (svanoort) [inline demo/repo/Jenkinsfile]: If that were any dirtier it would be banned as an obscenity... but it gets the job done and is better than the alternative.

**Comment by svanoort** [comment]
🐝  I like it and it subsumes some general and useful docker cleanup/formatting too.  

Might be worth adding a demo of silent steps (steps without stages) but if we are considering changing that we should hold off. 

Glad to see someone playing around with the nested stages -- I did a little of that for the block scoped stage demo.  They work mostly okay in stage view (aside from not showing them in parallel), due to all the work I put in with the new Bismuth modules.

**Comment by rsandell** [comment]
:bee:

File: .gitignore (hunk 5 above)
Comment (jglick) [inline .gitignore]: I will leave it to users of IDEs which require proprietary files to be added to Maven repositories to waste time added those ignores. :-)

**Comment by jglick** [comment]
> Might be worth adding a demo of silent steps

Nah, for the demo we want to show things we _recommend_, not everything that _could work_, and certainly not things which do not work at all.

**Pull request merged**